warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java

```diff
@@ -0,0 +1,223 @@
+import org.apache.accumulo.core.data.Key;
+import org.apache.commons.jexl2.parser.ASTFunctionNode;
+import org.apache.commons.jexl2.parser.ASTNotNode;
+import org.apache.commons.jexl2.parser.JexlNode;
+import org.apache.log4j.Logger;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * A function that intersects document keys prior to fetching term offsets.
+ * <p>
+ * This only works on positive phrase queries -- that is, no guarantee is made for negated phrases
+ */
+public class DocumentKeysFunction {
+    
+    private static final Logger log = Logger.getLogger(DocumentKeysFunction.class);
+
+    private final Set<String> negatedValues = new HashSet<>();
+    private final Multimap<String,ContentFunction> contentFunctions = LinkedListMultimap.create();
+    
+    /**
+     * @param config
+     *            a {@link TermFrequencyConfig}
+     */
+    public DocumentKeysFunction(TermFrequencyConfig config) {
+        populateContentFunctions(config.getScript());
+    }
+    
+    protected void populateContentFunctions(JexlNode node) {
+        
+        ContentFunctionsDescriptor descriptor = new ContentFunctionsDescriptor();
+        ContentJexlArgumentDescriptor argsDescriptor;
+        Set<String>[] fieldsAndTerms;
+        JexlNode parent;
+        String field;
+        
+        Multimap<String,Function> functions = TermOffsetPopulator.getContentFunctions(node);
+        for (String key : functions.keySet()) {
+            Collection<Function> coll = functions.get(key);
+            for (Function f : coll) {
+                parent = f.args().get(0).jjtGetParent();
+                
+                if (!(parent instanceof ASTFunctionNode)) {
+                    throw new IllegalArgumentException("parent was not a function node");
+                }
+                
+                argsDescriptor = descriptor.getArgumentDescriptor((ASTFunctionNode) parent);
+                
+                // content, tf, and indexed fields are not actually needed to extract fields from the function node
+                fieldsAndTerms = argsDescriptor.fieldsAndTerms(Collections.emptySet(), Collections.emptySet(), Collections.emptySet(), null);
+                
+                if (fieldsAndTerms[0].size() != 1) {
+                    throw new IllegalStateException("content function had more than one field");
+                }
+
+                field = JexlASTHelper.deconstructIdentifier(fieldsAndTerms[0].iterator().next());
+                ContentFunction contentFunction = new ContentFunction(field, fieldsAndTerms[1]);
+                contentFunctions.put(contentFunction.getField(), contentFunction);
+
+                if(isFunctionNegated(f)){
+                    negatedValues.addAll(contentFunction.getValues());
+                }
+            }
+        }
+    }
+
+    public boolean isFunctionNegated(Function f){
```

Comment:
At this point won't any nested negations already be flattened?

Reply:
https://github.com/NationalSecurityAgency/datawave/blob/integration/warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownNegationVisitor.java

I think as long as this is still being used we will always prevent having more than 1